# AI-ML Assignment – 1
## Topic: Medical Insurance Cost Prediction using Multiple Linear Regression

### Student Details
- **Name**: Arsh Baktoo  
- **Registration No**: 23BCE10430  
- **Application Number**: IN26010763  
- **Batch Number**: 2(B)  
- **Email ID**: arshbaktoo@gmail.com  
- **Submission Deadline**: 21 July 2026, 11:59 PM IST  

---

## Task 1: Data Understanding (2 Marks)

1. **Load the dataset using Pandas.**
2. **Display the first five records.**
3. **Identify:**
   - Numerical features
   - Categorical features
   - Target variable

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Set aesthetic visualization style
sns.set_theme(style="whitegrid")

In [2]:
# 1. Load the dataset using Pandas
df = pd.read_csv('insurance.csv')
print(f"Dataset successfully loaded. Total records: {df.shape[0]}, Total features: {df.shape[1]}")

Dataset successfully loaded. Total records: 1338, Total features: 7


In [3]:
# 2. Display the first five records
df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


In [4]:
# 3. Identify feature types and target variable
numerical_features = ['age', 'bmi', 'children']
categorical_features = ['sex', 'smoker', 'region']
target_variable = 'charges'

print("=== Identification of Variables ===")
print(f"Numerical Features   : {numerical_features}")
print(f"Categorical Features : {categorical_features}")
print(f"Target Variable      : {target_variable}")

=== Identification of Variables ===
Numerical Features   : ['age', 'bmi', 'children']
Categorical Features : ['sex', 'smoker', 'region']
Target Variable      : charges


--- 
## Task 2: Data Preprocessing (2 Marks)

Perform the following:
- Check for missing values.
- Encode categorical variables (`sex`, `smoker`, `region`).
- Split the dataset into 80% training and 20% testing.

In [5]:
# 1. Check for missing values
missing_values = df.isnull().sum()
print("Missing Values Count per Column:")
print(missing_values)
print(f"\nTotal missing values: {missing_values.sum()}")

Missing Values Count per Column:
age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

Total missing values: 0


In [6]:
# 2. Encode categorical variables using One-Hot Encoding (drop_first=True to avoid multicollinearity / dummy variable trap)
df_encoded = pd.get_dummies(df, columns=categorical_features, drop_first=True)
df_encoded.head()

,age,bmi,children,charges,sex_male,smoker_yes,region_northwest,region_southeast,region_southwest
0,19,27.900,0,16884.92400,False,True,False,False,True
1,18,33.770,1,1725.55230,True,False,False,True,False
2,28,33.000,3,4449.46200,True,False,False,True,False
3,33,22.705,0,21984.47061,True,False,True,False,False
4,32,28.880,0,3866.85520,True,False,True,False,False


In [7]:
# 3. Split the dataset into 80% training and 20% testing
X = df_encoded.drop(target_variable, axis=1)
y = df_encoded[target_variable]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Features (X) shape : {X.shape}")
print(f"Target (y) shape   : {y.shape}")
print(f"Training set shape : {X_train.shape} [80%]")
print(f"Testing set shape  : {X_test.shape} [20%]")

Features (X) shape : (1338, 8)
Target (y) shape   : (1338,)
Training set shape : (1070, 8) [80%]
Testing set shape  : (268, 8) [20%]


--- 
## Task 3: Model Development (3 Marks)

- Build a Multiple Linear Regression model using features (`age`, `sex`, `bmi`, `children`, `smoker`, `region`).
- Target Variable: `charges`.
- Train the model and predict the charges for the test dataset.

In [8]:
# Build and train the Multiple Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

print("Multiple Linear Regression model trained successfully.\n")
print(f"Model Intercept: {model.intercept_:.2f}\n")
print("Feature Coefficients:")
for col, coef in zip(X.columns, model.coef_):
    print(f"  {col:<20}: {coef:.2f}")

Multiple Linear Regression model trained successfully.

Model Intercept: -11931.22

Feature Coefficients:
  age                 : 256.98
  bmi                 : 337.09
  children            : 425.28
  sex_male            : -18.59
  smoker_yes          : 23651.13
  region_northwest    : -370.68
  region_southeast    : -657.86
  region_southwest    : -809.80


In [9]:
# Predict charges for the test dataset
y_pred = model.predict(X_test)

# Create comparison table for sample test cases
sample_comparison = pd.DataFrame({
    'Actual Charges ($)': y_test.values,
    'Predicted Charges ($)': y_pred,
    'Absolute Error ($)': np.abs(y_test.values - y_pred)
})
sample_comparison.head()

,Actual Charges ($),Predicted Charges ($),Absolute Error ($)
0,9095.06825,8924.41,170.66
1,5272.17580,7116.27,1844.09
2,29330.98315,36978.72,7647.74
3,9301.89355,12880.88,3578.99
4,33750.29180,27019.95,6730.34


--- 
## Task 4: Model Evaluation (2 Marks)

Evaluate the model using:
- Mean Absolute Error (MAE)
- Mean Squared Error (MSE)
- Root Mean Squared Error (RMSE)
- R2 Score ($R^2$)
- Actual vs Predicted Scatter Plot
- 2–3 Observations based on model performance.

In [10]:
# Compute Evaluation Metrics
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

print("=== Performance Metrics ===")
print(f"Mean Absolute Error (MAE) : ${mae:.2f}")
print(f"Mean Squared Error (MSE)  : {mse:.2f}")
print(f"Root Mean Squared Error   : ${rmse:.2f}")
print(f"R2 Score                  : {r2:.4f} ({r2*100:.2f}% variance explained)")

=== Performance Metrics ===
Mean Absolute Error (MAE) : $4181.19
Mean Squared Error (MSE)  : 33596915.85
Root Mean Squared Error   : $5796.28
R2 Score                  : 0.7836 (78.36% variance explained)


In [11]:
# Create Actual vs Predicted Scatter Plot
plt.figure(figsize=(9, 6), dpi=300)
plt.scatter(y_test, y_pred, alpha=0.75, color='#1f77b4', edgecolors='k', s=50, label='Predicted Points')

# Add reference line (y = x)
max_val = max(y_test.max(), y_pred.max())
min_val = min(y_test.min(), y_pred.min())
plt.plot([min_val, max_val], [min_val, max_val], 'r--', linewidth=2.5, label='Ideal Line (y = x)')

plt.title('Actual vs. Predicted Medical Insurance Charges', fontsize=14, fontweight='bold', pad=15)
plt.xlabel('Actual Charges ($)', fontsize=12)
plt.ylabel('Predicted Charges ($)', fontsize=12)
plt.legend(fontsize=11)
plt.tight_layout()
plt.savefig('actual_vs_predicted.png', dpi=300)
plt.show()

### Performance Observations

1. **Strong Explanatory Power ($R^2 = 0.7836$):** The Multiple Linear Regression model successfully accounts for approximately **78.36%** of the total variance in medical insurance charges using the provided features.
2. **Dominance of Smoking Status:** The feature coefficient for `smoker_yes` is approximately **+$23,651.13**, proving that smoking is overwhelmingly the largest contributor to elevated insurance costs compared to age (+$256.98/year) and BMI (+$337.09/unit).
3. **Heteroscedasticity at Extreme Values:** The scatter plot shows increased spread and underestimation for actual charges exceeding $30,000. This indicates non-linear interaction effects (such as high BMI combined with smoking) that standard linear regression cannot capture without non-linear terms.

--- 
## Task 5: Conclusion (1 Mark)

Write a 100–150 word conclusion covering:
- Key findings
- Factors affecting insurance charges
- One limitation of Linear Regression for this problem

### Conclusion

This project developed a Multiple Linear Regression model to estimate medical insurance charges based on customer demographic and health attributes. Key findings demonstrate that lifestyle and physical health metrics are primary drivers of insurance expenditure. Specifically, smoking status exerts the most dominant positive impact, increasing predicted charges by approximately $23,651, followed by age (+$257 per year) and BMI (+$337 per unit). The model achieved an $R^2$ score of 0.7836 and a Mean Absolute Error of $4,181.19. A major limitation of Multiple Linear Regression in this context is its strict assumption of linear, additive relationships. In reality, health risk factors exhibit strong non-linear interactions—such as the compounding effect between high BMI and smoking—which leads to underestimation of extreme charges for high-risk individuals.